helix-term/src/commands/engine/scheme.rs

```diff
@@ -0,0 +1,2003 @@
+        };
+
+        module.register_fn("set-options", func);
+    }
+
+    module.register_value(
+        "PromptEvent::Validate",
+        PromptEvent::Validate.into_steelval().unwrap(),
+    );
+    module.register_value(
+        "PromptEvent::Update",
+        PromptEvent::Update.into_steelval().unwrap(),
+    );
+
+    // Register everything in the typable command list. Now these are all available
+    for command in TYPABLE_COMMAND_LIST {
+        let func = |cx: &mut Context, args: &[Cow<str>], event: PromptEvent| {
+            let mut cx = compositor::Context {
+                editor: cx.editor,
+                scroll: None,
+                jobs: cx.jobs,
+            };
+
+            (command.fun)(&mut cx, args, event)
+        };
+
+        module.register_fn(command.name, func);
+    }
+
+    engine.register_module(module);
+}
+
+fn load_editor_api(engine: &mut Engine, _api: EditorApi) {
+    let mut module = BuiltInModule::new("helix/core/editor");
+
+    RegisterFn::<
+        _,
+        steel::steel_vm::register_fn::MarkerWrapper7<(
+            Context<'_>,
+            helix_view::Editor,
+            helix_view::Editor,
+            Context<'static>,
+        )>,
+        helix_view::Editor,
+    >::register_fn(&mut module, "cx-editor!", get_editor);
+
+    module.register_fn("set-scratch-buffer-name!", set_scratch_buffer_name);
+
+    module.register_fn("editor-focus", current_focus);
+    module.register_fn("editor->doc-id", get_document_id);
+    module.register_fn("doc-id->usize", document_id_to_usize);
+    module.register_fn("editor-switch!", switch);
+    module.register_fn("editor-set-focus!", Editor::focus);
+    module.register_fn("editor-mode", editor_get_mode);
+    module.register_fn("editor-set-mode!", editor_set_mode);
+    module.register_fn("editor-doc-in-view?", is_document_in_view);
+
+    // TODO: These are some horrendous type annotations, however... they do work.
+    // If the type annotations are a bit more ergonomic, we might be able to get away with this
+    // (i.e. if they're sensible enough)
+    RegisterFn::<
+        _,
+        steel::steel_vm::register_fn::MarkerWrapper8<(
+            helix_view::Editor,
+            DocumentId,
+            Document,
+            Document,
+            helix_view::Editor,
+        )>,
+        Document,
+    >::register_fn(&mut module, "editor->get-document", get_document);
+
+    // Check if the doc exists first
+    module.register_fn("editor-doc-exists?", document_exists);
+    module.register_fn("Document-path", document_path);
+    module.register_fn("Document-focused-at", document_focused_at);
+    module.register_fn("editor-all-documents", editor_all_documents);
+
+    module.register_fn("helix.context?", is_context);
+    module.register_type::<DocumentId>("DocumentId?");
+
+    module.register_fn("editor-cursor", Editor::cursor);
+
+    module.register_fn("cx->cursor", |cx: &mut Context| cx.editor.cursor());
```

Comment:
this is an instance of exposing too many implementation details.

As I said in the other comment, context should be somekind of hidden global (which can have different capabilities depending on callsite).

The user should only know that he can do X/Y at a certain callsite. the existance of context and editor struct is an implementation detail that plugins should not be aware of

Reply:
So just to clarify - don't expose the editor struct at all, any details go directly through the context?

I'm fine with that, just confirming the intent here.

Reply:
yeah don't expose theses structs at all. These should not be `editor-cursor` and `cx->cursor` but instead just `active-cursor` and the internal structs are handled in the background.